File: cwclock-mobile/src/screens/organizations/OrganizationScreen.js
import React, { useEffect, useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import Screen from "../../components/Screen";
import FormField from "../../components/FormField";
import SelectField from "../../components/SelectField";
import Button from "../../components/Button";
import ErrorBanner from "../../components/ErrorBanner";
import { useI18n } from "../../i18n/I18nContext";
import { apiErrorMessage } from "../../i18n/translate";
import { isOrgOwner } from "../../common/permissions";
import { listCountriesApi } from "../../redux/countries/countries.actions";
import { listCurrenciesApi } from "../../redux/currencies/currencies.actions";
import { updateOrganizationApi } from "../../redux/organizations/organizations.actions";

// OrganizationScreen: view/edit the current organization's profile. Ported
// from cwclock-ui's Organizations.jsx edit form, trimmed to the core fields
// (no logo/stamp upload - no image-picker dependency in this app - and no
// external-connections management, both left to the web app). Editing is
// owner-only, matching OrganizationHandler.Update's route gate
// (RequireRole(RoleOwner)) - other roles see a read-only view.
const OrganizationScreen = () => {
  const { t, locale } = useI18n();
  const dispatch = useDispatch();
  const { orgId, user } = useSelector((state) => state.session);
  const { items: orgs } = useSelector((state) => state.organizations);
  const { items: countries } = useSelector((state) => state.countries);
  const { items: currencies } = useSelector((state) => state.currencies);

  const org = orgs.find((o) => o.id === orgId);
  const canEdit = isOrgOwner(user, org);

  const [fields, setFields] = useState(null);
  const [error, setError] = useState(null);
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    dispatch(listCountriesApi());
    dispatch(listCurrenciesApi());
  }, [dispatch]);

  useEffect(() => {
    if (org && !fields) {
      setFields({
        name: org.name || "",
        accountingEmail: org.accountingEmail || "",
        address: org.address || "",
        postalCode: org.postalCode || "",
        city: org.city || "",
        country: org.country || "",
        currency: org.currency || "",
        vatNumber: org.vatNumber || "",
        identificationNumber: org.identificationNumber || "",
        iban: org.iban || "",
        bic: org.bic || "",
      });
    }
  }, [org, fields]);

  const countryItems = useMemo(() => countries.map((c) => ({ value: c.iso, label: c.name })), [countries]);
  const currencyItems = useMemo(() => currencies.map((c) => ({ value: c.iso, label: c.name })), [currencies]);

  if (!org || !fields) return null;

  const setField = (key, value) => setFields({ ...fields, [key]: value });

  const handleSave = async () => {
    setError(null);
    setSaving(true);
    try {
      await dispatch(updateOrganizationApi(orgId, fields));
    } catch (e) {
      setError(apiErrorMessage(e, locale));
    } finally {
      setSaving(false);
    }
  };

  return (
    <Screen>
      {!canEdit && <ErrorBanner message={t("organizations.ownerOnlyHint")} />}
      <FormField
        label={t("organizations.name")}
        value={fields.name}
        onChangeText={(v) => setField("name", v)}
        editable={canEdit}
      />
      <FormField
        label={t("organizations.accountingEmail")}
        value={fields.accountingEmail}
        onChangeText={(v) => setField("accountingEmail", v)}
        autoCapitalize="none"
        keyboardType="email-address"
        editable={canEdit}
      />
      <FormField
        label={t("organizations.address")}
        value={fields.address}
        onChangeText={(v) => setField("address", v)}
        editable={canEdit}
      />
      <FormField
        label={t("organizations.postalCode")}
        value={fields.postalCode}
        onChangeText={(v) => setField("postalCode", v)}
        editable={canEdit}
      />
      <FormField
        label={t("organizations.city")}
        value={fields.city}
        onChangeText={(v) => setField("city", v)}
        editable={canEdit}
      />
      <SelectField
        label={t("organizations.country")}
        value={fields.country}
        onValueChange={(v) => setField("country", v)}
        items={countryItems}
        placeholder={t("organizations.country")}
        enabled={canEdit}
      />
      <SelectField
        label={t("organizations.currency")}
        value={fields.currency}
        onValueChange={(v) => setField("currency", v)}
        items={currencyItems}
        placeholder={t("organizations.currency")}
        enabled={canEdit}
      />
      <FormField
        label={t("organizations.vatNumber")}
        value={fields.vatNumber}
        onChangeText={(v) => setField("vatNumber", v)}
        editable={canEdit}
      />
      <FormField
        label={t("organizations.identificationNumber")}
        value={fields.identificationNumber}
        onChangeText={(v) => setField("identificationNumber", v)}
        editable={canEdit}
      />
      <FormField
        label={t("organizations.iban")}
        value={fields.iban}
        onChangeText={(v) => setField("iban", v)}
        autoCapitalize="characters"
        editable={canEdit}
      />
      <FormField
        label={t("organizations.bic")}
        value={fields.bic}
        onChangeText={(v) => setField("bic", v)}
        autoCapitalize="characters"
        editable={canEdit}
      />
      <ErrorBanner message={error} />
      {canEdit && <Button title={t("common.save")} onPress={handleSave} loading={saving} />}
    </Screen>
  );
};

export default OrganizationScreen;

File: cwclock-mobile/lib/screens/organizations/organization_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../api/api_client.dart';
import '../../i18n/app_localizations.dart';
import '../../providers/countries_provider.dart';
import '../../providers/currencies_provider.dart';
import '../../providers/locale_provider.dart';
import '../../providers/organizations_provider.dart';
import '../../providers/permissions.dart' as perm;
import '../../providers/session_provider.dart';
import '../../widgets/app_button.dart';
import '../../widgets/app_screen.dart';
import '../../widgets/error_banner.dart';
import '../../widgets/form_field.dart';
import '../../widgets/select_field.dart';

/// View/edit the current organization's profile. Editing is owner-only,
/// matching OrganizationHandler.Update's route gate - other roles see a
/// read-only view. Ported from
/// src/screens/organizations/OrganizationScreen.js.
class OrganizationScreen extends ConsumerStatefulWidget {
  const OrganizationScreen({super.key});

  @override
  ConsumerState<OrganizationScreen> createState() => _OrganizationScreenState();
}

class _OrganizationScreenState extends ConsumerState<OrganizationScreen> {
  Map<String, String>? _fields;
  String? _error;
  bool _saving = false;

  @override
  void initState() {
    super.initState();
    Future.microtask(() {
      ref.read(countriesProvider.notifier).listCountries();
      ref.read(currenciesProvider.notifier).listCurrencies();
    });
  }

  void _setField(String key, String value) => setState(() => _fields = {...?_fields, key: value});

  Future<void> _handleSave() async {
    final orgId = ref.read(sessionProvider).orgId;
    if (orgId == null || _fields == null) return;
    final locale = ref.read(localeProvider);
    setState(() {
      _error = null;
      _saving = true;
    });
    try {
      await ref.read(organizationsProvider.notifier).updateOrganization(orgId, _fields!);
    } catch (e) {
      setState(() => _error = apiErrorMessage(asApiException(e), locale));
    } finally {
      if (mounted) setState(() => _saving = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    final locale = ref.watch(localeProvider);
    final t = translateWith(locale);
    final session = ref.watch(sessionProvider);
    final orgsState = ref.watch(organizationsProvider);
    final countries = ref.watch(countriesProvider);
    final currencies = ref.watch(currenciesProvider);

    final org = orgsState.items.where((o) => o.id == session.orgId).firstOrNull;
    if (org == null) return const SizedBox.shrink();

    _fields ??= {
      'name': org.name,
      'accountingEmail': org.accountingEmail ?? '',
      'address': org.address ?? '',
      'postalCode': org.postalCode ?? '',
      'city': org.city ?? '',
      'country': org.country ?? '',
      'currency': org.currency ?? '',
      'vatNumber': org.vatNumber ?? '',
      'identificationNumber': org.identificationNumber ?? '',
      'iban': org.iban ?? '',
      'bic': org.bic ?? '',
    };
    final fields = _fields!;
    final canEdit = perm.isOrgOwner(session.user, org);

    final countryItems = [for (final c in countries) SelectItem(c.iso, c.name)];
    final currencyItems = [for (final c in currencies) SelectItem(c.iso, c.name)];

    return Scaffold(
      appBar: AppBar(title: Text(t('organizations.profileTitle'))),
      body: AppScreen(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            if (!canEdit) ErrorBanner(message: t('organizations.ownerOnlyHint')),
            AppFormField(
              label: t('organizations.name'),
              value: fields['name']!,
              onChanged: (v) => _setField('name', v),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.accountingEmail'),
              value: fields['accountingEmail']!,
              onChanged: (v) => _setField('accountingEmail', v),
              enabled: canEdit,
              keyboardType: TextInputType.emailAddress,
              textCapitalization: TextCapitalization.none,
            ),
            AppFormField(
              label: t('organizations.address'),
              value: fields['address']!,
              onChanged: (v) => _setField('address', v),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.postalCode'),
              value: fields['postalCode']!,
              onChanged: (v) => _setField('postalCode', v),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.city'),
              value: fields['city']!,
              onChanged: (v) => _setField('city', v),
              enabled: canEdit,
            ),
            AppSelectField(
              label: t('organizations.country'),
              value: fields['country']!,
              onChanged: (v) => _setField('country', v),
              items: countryItems,
              placeholder: t('organizations.country'),
              enabled: canEdit,
            ),
            AppSelectField(
              label: t('organizations.currency'),
              value: fields['currency']!,
              onChanged: (v) => _setField('currency', v),
              items: currencyItems,
              placeholder: t('organizations.currency'),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.vatNumber'),
              value: fields['vatNumber']!,
              onChanged: (v) => _setField('vatNumber', v),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.identificationNumber'),
              value: fields['identificationNumber']!,
              onChanged: (v) => _setField('identificationNumber', v),
              enabled: canEdit,
            ),
            AppFormField(
              label: t('organizations.iban'),
              value: fields['iban']!,
              onChanged: (v) => _setField('iban', v),
              enabled: canEdit,
              textCapitalization: TextCapitalization.characters,
            ),
            AppFormField(
              label: t('organizations.bic'),
              value: fields['bic']!,
              onChanged: (v) => _setField('bic', v),
              enabled: canEdit,
              textCapitalization: TextCapitalization.characters,
            ),
            ErrorBanner(message: _error),
            if (canEdit) AppButton(title: t('common.save'), onPressed: _handleSave, loading: _saving),
          ],
        ),
      ),
    );
  }
}

extension _FirstOrNull<T> on Iterable<T> {
  T? get firstOrNull {
    final it = iterator;
    return it.moveNext() ? it.current : null;
  }
}
